File: ros2_ws/src/469labex/le1_py_pkg/le1_py_pkg/cumulative_adder.py
#!/usr/bin/env python3
import rclpy
from rclpy.lifecycle import LifecycleNode
from rclpy.lifecycle.node import LifecycleState, TransitionCallbackReturn
from example_interfaces.msg import Int64

class CumulativeAdderNode(LifecycleNode):
    def __init__(self):
        super().__init__("cumulative_adder")
        # plain instance variables go here — e.g. wherever you track the running sum
        # NOT the publisher or subscriber

    def on_configure(self, previous_state: LifecycleState):
        # create the publisher here (create_lifecycle_publisher, not create_publisher)
        # create the subscriber here (create_subscription — same call as before, just moved)
        # return TransitionCallbackReturn.SUCCESS

    def on_activate(self, previous_state: LifecycleState):
        # same one-liner pattern as my_talker.py's on_activate

    def on_deactivate(self, previous_state: LifecycleState):
        # same one-liner pattern as my_talker.py's on_deactivate

    def on_cleanup(self, previous_state: LifecycleState):
        # destroy the publisher and subscriber you created in on_configure
        # return TransitionCallbackReturn.SUCCESS

    def on_shutdown(self, previous_state: LifecycleState):
        # same cleanup as on_cleanup
        # return TransitionCallbackReturn.SUCCESS

    def callback_integer(self, msg):
        # keep your original logic here — add to the sum, publish, log
        pass

def main(args=None):
    # follow the same structure as my_talker.py's main()
    pass

if __name__ == "__main__":
    main()

File: ros2_ws/src/469labex/le4_cpp_pkg/src/cumulative_adder.cpp
#include "rclcpp/rclcpp.hpp"
#include "example_interfaces/msg/int64.hpp"

class CumulativeAdderNode : public rclcpp::Node
{
public:
  CumulativeAdderNode() : Node("cumulative_adder"), sum_(0)
  {
    subscriber_ = create_subscription<example_interfaces::msg::Int64>(
        "integer", 10,
        [this](example_interfaces::msg::Int64::SharedPtr msg)
        { callback_integer(msg); });
    publisher_ = create_publisher<example_interfaces::msg::Int64>("integer_sum", 10);
    RCLCPP_INFO(get_logger(), "The cumulative adder has started.");
  }

private:
  void callback_integer(const example_interfaces::msg::Int64::SharedPtr msg)
  {
    sum_ += msg->data;
    auto sum_msg = example_interfaces::msg::Int64();
    sum_msg.data = sum_;
    publisher_->publish(sum_msg);
  }

  int64_t sum_;
  rclcpp::Subscription<example_interfaces::msg::Int64>::SharedPtr subscriber_;
  rclcpp::Publisher<example_interfaces::msg::Int64>::SharedPtr publisher_;
};

int main(int argc, char **argv)
{
  rclcpp::init(argc, argv);
  auto node = std::make_shared<CumulativeAdderNode>();
  rclcpp::spin(node);
  rclcpp::shutdown();
  return 0;
}